File: next/dweb-browser/MicroService/src/Sys/Dns/DnsNMM.cs
using DwebBrowser.MicroService.Sys.Http;

namespace DwebBrowser.MicroService.Sys.Dns;

public class DnsNMM : NativeMicroModule
{
    static readonly Debugger Console = new("DnsNMM");
    // 已安装的应用
    private Dictionary<Mmid, MicroModule> _installApps = new();

    // 正在运行的应用
    private Dictionary<Mmid, PromiseOut<MicroModule>> _runningApps = new();
    public override List<Dweb_DeepLink> Dweb_deeplinks { get; init; } = new() { "dweb:open" };
    public override List<MicroModuleCategory> Categories { get; init; } = new()
    {
        MicroModuleCategory.Service,
        MicroModuleCategory.Routing_Service,
    };

    public new const string Name = "Dweb Name System";
    public override string? ShortName { get; set; } = "DNS";
    public DnsNMM() : base("dns.sys.dweb")
    {
    }

    public async Task Bootstrap()
    {
        if (!Running)
        {
            await BootstrapMicroModule(this);
        }

        await OnActivity();
    }

    public Task BootstrapMicroModule(MicroModule fromMM) =>
        fromMM.BootstrapAsync(new MyBootstrapContext(new MyDnsMicroModule(this, fromMM))).NoThrow();

    public class MyBootstrapContext : IBootstrapContext
    {
        public IDnsMicroModule Dns { get; init; }

        public MyBootstrapContext(MyDnsMicroModule dns)
        {
            Dns = dns;
        }
    }

    record MM(Mmid fromMmid, Mmid toMmid)
    {
        public static Dictionary<Mmid, Dictionary<Mmid, MM>> Values = new();
        public static MM From(Mmid fromMmid, Mmid toMmid) =>
            Values.GetValueOrPut(fromMmid, () => new()).GetValueOrPut(toMmid, () => new MM(fromMmid, toMmid));
    }

    /** <summary>对等连接列表</summary> */
    private Dictionary<MM, PromiseOut<ConnectResult>> _mmConnectsMap = new();

    /** <summary>为两个mm建立 ipc 通讯</summary> */
    private Task<ConnectResult> _connectTo(MicroModule fromMM, Mmid toMmid, PureRequest reason)
    {
        PromiseOut<ConnectResult> wait;
        lock (_mmConnectsMap)
        {
            var mmKey = MM.From(fromMM.Mmid, toMmid);

            /**
             * 一个互联实例
             */
            wait = _mmConnectsMap.GetValueOrPut(mmKey, () =>
            {
                return new PromiseOut<ConnectResult>().Also(async po =>
                {
                    Console.Log("ConnectTo", "DNS/opening {0} => {1}", fromMM.Mmid, toMmid);
                    var toMM = await Open(toMmid);
                    Console.Log("ConnectTo", "DNS/connect {0} => {1}", fromMM.Mmid, toMmid);
                    var connectResult = await NativeConnect.ConnectMicroModulesAsync(fromMM, toMM, reason);
                    connectResult.IpcForFromMM.OnClose += async (_) =>
                    {
                        lock (_mmConnectsMap)
                        {
                            _mmConnectsMap.Remove(mmKey);
                        }
                    };
                    po.Resolve(connectResult);

                    /// 如果可以，反向存储
                    if (connectResult.IpcForToMM is not null)
                    {
                        var mmkey2 = MM.From(toMmid, fromMM.Mmid);
                        lock (_mmConnectsMap)
                        {
                            _mmConnectsMap.GetValueOrPut(mmkey2, () =>
                            {
                                return new PromiseOut<ConnectResult>().Also(po2 =>
                                {
                                    var connectResult2 = new ConnectResult(
                                    connectResult.IpcForToMM, connectResult.IpcForFromMM);

                                    connectResult2.IpcForFromMM.OnClose += async (_) =>
                                    {
                                        lock (_mmConnectsMap)
                                        {
                                            _mmConnectsMap.Remove(mmkey2);
                                        }
                                    };
                                    po2.Resolve(connectResult2);
                                });
                            });
                        }
                    }
                });
            });

            Console.Log("_connectTo", "new MM => {0}", _mmConnectsMap.GetValueOrDefault(new MM(fromMM.Mmid, toMmid)));
        }

        return wait.WaitPromiseAsync();
    }

    public class MyDnsMicroModule : IDnsMicroModule
    {
        private DnsNMM _dnsMM { get; init; }
        private MicroModule _fromMM { get; init; }

        public MyDnsMicroModule(DnsNMM dnsMM, MicroModule fromMM)
        {
            _dnsMM = dnsMM;
            _fromMM = fromMM;
        }

        public void Install(MicroModule mm)
        {
            _dnsMM.Install(mm);
        }

        public bool UnInstall(MicroModule mm)
        {
            return _dnsMM.UnInstall(mm);
        }

        public async Task<IMicroModuleManifest?> Query(Mmid mmid) =>
            (await _dnsMM.Query(mmid)) is MicroModule mm ? mm : null;

        public async Task Restart(Mmid mmid)
        {
            // 关闭后端连接
            await _dnsMM.Close(mmid);
            await _dnsMM.Open(mmid);
        }

        public Task<ConnectResult> ConnectAsync(Mmid mmid, PureRequest? reason = null)
        {
            return _dnsMM._connectTo(
                _fromMM, mmid, reason ?? new PureRequest(string.Format("file://{0}", mmid), IpcMethod.Get));
        }

        public async Task<bool> Open(Mmid mmid)
        {
            /// 已经在运行中了，直接返回true
            if (_dnsMM._runningApps.ContainsKey(mmid))
            {
                return true;
            }

            /// 尝试运行，成功就返回true
            try
            {
                await _dnsMM.Open(mmid);
                return true;
            }
            catch
            {
                return false;
            }
        }

        public async Task<bool> Close(Mmid mmid)
        {
            if (_dnsMM._runningApps.ContainsKey(mmid))
            {
                if (await _dnsMM.Close(mmid) == 1)
                {
                    return true;
                }
            }

            return false;
        }

        public async Task<MicroModule[]> Search(MicroModuleCategory category)
        {
            return _dnsMM.Search(category).ToArray();
        }
    }

    protected override async Task _bootstrapAsync(IBootstrapContext bootstrapContext)
    {
        Install(this);
        _runningApps.Add(Mmid, PromiseOut<MicroModule>.StaticResolve(this));

        /**
         * 对全局的自定义路由提供适配器
         * 对 nativeFetch 定义 file://xxx.dweb的解析
         */
        var cb = NativeFetch.NativeFetchAdaptersManager.Append(async (fromMM, request) =>
        {
            if (request.ParsedUrl is not null and var parsedUrl && parsedUrl.Scheme is "file" && parsedUrl.Hostname is var mmid && mmid.EndsWith(".dweb"))
            {
                var microModule = _installApps.GetValueOrDefault(mmid);

                if (microModule is not null)
                {
                    var connectResult = await _connectTo(fromMM, mmid, request);
                    Console.Log("NativeFetch", "DNS/request/{0} => {1} [{2}] {3}", fromMM.Mmid, mmid, request.Method.Method, parsedUrl.Path);
                    return await connectResult.IpcForFromMM.Request(request);
                }

                return new PureResponse(HttpStatusCode.BadGateway, Body: new PureUtf8StringBody(request.Url));
            }

            return null;
        });

        /// dweb-deeplinks
        var deeplinkCb = NativeFetch.NativeFetchAdaptersManager.Append(async (fromMM, request) =>
        {
            if (request.ParsedUrl is not null and var parsedUrl && parsedUrl.Scheme is "dweb" && parsedUrl.Hostname == "")
            {
                Console.Log("NativeFetch", "DNS/dwebDeepLinks path={0}, host={1}", parsedUrl.Path, parsedUrl.Hostname);

                foreach (var app in _installApps.Values)
                {
                    if (app.Dweb_deeplinks.Contains(string.Format("dweb:{0}", parsedUrl.Path)))
                    {
                        /// 加上///的原因：不修改url的话，不符合url标准，会异常，加上///可以进行路由匹配
                        request.Url = request.Url.Replace("dweb:", "dweb:///");
                        var connectResult = await _connectTo(fromMM, app.Mmid, request);
                        return await connectResult.IpcForFromMM.Request(request);
                    }
                }

                return new PureResponse(HttpStatusCode.BadGateway, Body: new PureUtf8StringBody(request.Url));
            }

            return null;
        });

        OnAfterShutdown += async (_) => { cb(); deeplinkCb(); };

        var Query_appId = (URL parsedUrl) => parsedUrl.SearchParams.Get("app_id") ?? throw new ArgumentException("no found app_id");

        // 打开应用
        HttpRouter.AddRoute(IpcMethod.Get, "/open", async (request, _) =>
        {
            var parsedUrl = request.SafeUrl;
            Console.Log("Open", "{0} {1}", Mmid, parsedUrl.Path);
            await Open(parsedUrl.SearchParams.ForceGet("app_id"));
            return true;
        });

        // 关闭应用
        // TODO 能否关闭一个应该应该由应用自己决定
        HttpRouter.AddRoute(IpcMethod.Get, "/close", async (request, _) =>
        {
            var parsedUrl = request.SafeUrl;
            Console.Log("Close", "{0} {1}", Mmid, parsedUrl.Path);
            await Close(parsedUrl.SearchParams.ForceGet("app_id"));
            return true;
        });

        // deeplink
        HttpRouter.AddRoute(new Gateway.RouteConfig("open/", IpcMethod.Get), async (request, _) =>
        {
            var parsedUrl = request.SafeUrl;
            var app_id = parsedUrl.Path.Replace("open/", "");
            await Open(app_id);
            return true;
        });
    }

    protected override Task _onActivityAsync(IpcEvent Event, Ipc ipc) => OnActivity(Event);

    public async Task OnActivity(IpcEvent? Event = null)
    {
        Event ??= IpcEvent.FromUtf8("activity", "");

        /// 启动 boot 模块
        await Open("boot.sys.dweb");
        var ipc = await ConnectAsync("boot.sys.dweb");

        if (ipc is not null)
        {
            await ipc.PostMessageAsync(Event);
        }
    }

    protected override async Task _shutdownAsync()
    {
        foreach (var mm in _installApps)
        {
            await mm.Value.ShutdownAsync();
        }

        _installApps.Clear();
    }

    /** <summary>安装应用</summary> */
    public void Install(MicroModule mm) => _installApps.TryAdd(mm.Mmid, mm);

    /** <summary>卸载应用</summary> */
    public bool UnInstall(MicroModule mm) => _installApps.Remove(mm.Mmid);

    /** <summary>查询应用</summary> */
    public async Task<MicroModule?> Query(Mmid mmid) => _installApps.GetValueOrDefault(mmid);

    public IEnumerable<MicroModule> Search(MicroModuleCategory category)
    {
        foreach (var app in _installApps.Values)
        {
            if (app.Categories.Contains(category))
            {
                yield return app;
            }
        }
    }

    /** <summary>打开应用</summary> */
    private Task<PromiseOut<MicroModule>> _Open(Mmid mmid)
    {
        return _runningApps.GetValueOrPutAsync(mmid, async () =>
        {
            var po = new PromiseOut<MicroModule>();

            var openingMM = await Query(mmid);
            _ = Task.Run(async () =>
            {
                if (openingMM is not null)
                {
                    await BootstrapMicroModule(openingMM);
                    openingMM.OnAfterShutdown += async (_) =>
                    {
                        await _removeRunningApps(mmid);
                    };
                    po.Resolve(openingMM);
                }
            }).NoThrow();

            return po;
        });
    }
    /** <summary>打开应用</summary> */
    public async Task<MicroModule> Open(Mmid mmid) => await (await _Open(mmid)).WaitPromiseAsync();

    /** <summary>关闭应用</summary> */
    public async Task<int> Close(Mmid mmid)
    {
        var microModule = await _removeRunningApps(mmid);

        if (microModule is null)
        {
            return -1;
        }

        try
        {
            await microModule.ShutdownAsync();

            return 1;
        }
        catch (Exception e)
        {
            Console.Warn("Close", "exception: {0}", e.Message);
            return 0;
        }
    }

    private async Task<MicroModule?> _removeRunningApps(Mmid mmid)
    {
        var microModulePo = _runningApps.GetValueOrDefault(mmid);

        if (microModulePo is null)
        {
            return null;
        }

        var app = await microModulePo.WaitPromiseAsync();

        if (app is not null)
        {
            _runningApps.Remove(mmid);
        }

        return app;
    }
}

File: next/kmp/microService/src/androidMain/kotlin/org/dweb_browser/microservice/sys/dns/DnsNMM.kt
package org.dweb_browser.microservice.sys.dns

import kotlinx.coroutines.launch
import kotlinx.coroutines.sync.Mutex
import kotlinx.coroutines.sync.withLock
import kotlinx.serialization.encodeToString
import kotlinx.serialization.json.Json
import org.dweb_browser.helper.ChangeState
import org.dweb_browser.helper.ChangeableMap
import org.dweb_browser.helper.PromiseOut
import org.dweb_browser.helper.printDebug
import org.dweb_browser.helper.removeWhen
import org.dweb_browser.helper.toJsonElement
import org.dweb_browser.microservice.core.BootstrapContext
import org.dweb_browser.microservice.core.ConnectResult
import org.dweb_browser.microservice.core.DnsMicroModule
import org.dweb_browser.microservice.core.MicroModule
import org.dweb_browser.microservice.core.NativeMicroModule
import org.dweb_browser.microservice.core.connectMicroModules
import org.dweb_browser.microservice.help.InitRequest
import org.dweb_browser.microservice.help.buildRequestX
import org.dweb_browser.microservice.help.types.MICRO_MODULE_CATEGORY
import org.dweb_browser.microservice.help.types.MMID
import org.dweb_browser.microservice.ipc.helper.IpcEvent
import org.dweb_browser.microservice.ipc.helper.ReadableStream
import org.http4k.core.Method
import org.http4k.core.Request
import org.http4k.core.Response
import org.http4k.core.Status
import org.http4k.core.Uri
import org.http4k.lens.Query
import org.http4k.lens.string
import org.http4k.routing.bind
import org.http4k.routing.routes

fun debugDNS(tag: String, msg: Any = "", err: Throwable? = null) =
  printDebug("fetch", tag, msg, err)

class DnsNMM : NativeMicroModule("dns.std.dweb", "Dweb Name System") {
  init {

    dweb_deeplinks = mutableListOf("dweb:open")
    short_name = "DNS";
    categories = listOf(MICRO_MODULE_CATEGORY.Service, MICRO_MODULE_CATEGORY.Routing_Service);
  }

  private val installApps = ChangeableMap<MMID, MicroModule>() // 已安装的应用
  private val runningApps = ChangeableMap<MMID, PromiseOut<MicroModule>>() // 正在运行的应用

  suspend fun bootstrap() {
    if (!this.running) {
      bootstrapMicroModule(this)
    }
  }

  data class MM(val fromMMID: MMID, val toMMID: MMID) {
    companion object {
      val values = mutableMapOf<MMID, MutableMap<MMID, MM>>()
      fun from(fromMMID: MMID, toMMID: MMID) =
        values.getOrPut(fromMMID) { mutableMapOf() }.getOrPut(toMMID) { MM(fromMMID, toMMID) }
    }
  }

  /** 对等连接列表 */
  private val mmConnectsMap = mutableMapOf<MM, PromiseOut<ConnectResult>>()
  private val mmConnectsMapLock = Mutex()


  /** 为两个mm建立 ipc 通讯 */
  private suspend fun connectTo(
    fromMM: MicroModule, toMMID: MMID, reason: Request
  ) = mmConnectsMapLock.withLock {
    val mmKey = MM.from(fromMM.mmid, toMMID)
    /**
     * 一个互联实例
     */
    mmConnectsMap.getOrPut(mmKey) {
      PromiseOut<ConnectResult>().alsoLaunchIn(ioAsyncScope) {
        debugFetch("DNS/open", "${fromMM.mmid} => $toMMID")
        val toMM = open(toMMID)
        debugFetch("DNS/connect", "${fromMM.mmid} <=> $toMMID")
        val connectResult = connectMicroModules(fromMM, toMM, reason)
        connectResult.ipcForFromMM.onClose {
          mmConnectsMap.remove(mmKey)
        }

        // 如果可以，反向存储
        if (connectResult.ipcForToMM != null) {
          val mmKey2 = MM.from(toMMID, fromMM.mmid)
          mmConnectsMapLock.withLock {
            mmConnectsMap.getOrPut(mmKey2) {
              PromiseOut<ConnectResult>().also { po2 ->
                val connectResult2 = ConnectResult(
                  connectResult.ipcForToMM, connectResult.ipcForFromMM
                );
                connectResult2.ipcForToMM?.also {
                  it.onClose {
                    mmConnectsMap.remove(mmKey2)
                  }
                }
                po2.resolve(connectResult2)
              }
            }
          }
        }
        connectResult
      }
    }

  }.waitPromise()

  class MyDnsMicroModule(private val dnsMM: DnsNMM, private val fromMM: MicroModule) :
    DnsMicroModule {
    override fun install(mm: MicroModule) {
      // TODO 作用域保护
      dnsMM.install(mm)
    }

    override fun uninstall(mmid: MMID): Boolean {
      // TODO 作用域保护
      return dnsMM.uninstall(mmid)
    }

    override fun query(mmid: MMID): MicroModule? {
      return dnsMM.query(mmid)
    }

    override suspend fun search(category: MICRO_MODULE_CATEGORY): MutableList<MicroModule> {
      return dnsMM.search(category)
    }

    // 调用重启
    override suspend fun restart(mmid: MMID) {
      // 关闭后端连接
      dnsMM.close(mmid)
      dnsMM.open(mmid)
    }

    override suspend fun connect(
      mmid: MMID, reason: Request?
    ): ConnectResult {
      // TODO 权限保护
      return dnsMM.connectTo(
        fromMM, mmid, reason ?: Request(Method.GET, Uri.of("file://$mmid"))
      )
    }

    override suspend fun open(mmid: MMID): Boolean {
      if (this.dnsMM.runningApps[mmid] == null) {
        return false
      }
      dnsMM.open(mmid)
      return true
    }

    override suspend fun close(mmid: MMID): Boolean {
      if (this.dnsMM.runningApps[mmid] !== null) {
        dnsMM.close(mmid);
        return true;
      }
      return false;
    }
  }

  class MyBootstrapContext(override val dns: MyDnsMicroModule) : BootstrapContext {}

  suspend fun bootstrapMicroModule(fromMM: MicroModule) {
    fromMM.bootstrap(MyBootstrapContext(MyDnsMicroModule(this@DnsNMM, fromMM)))
  }

  override suspend fun _bootstrap(bootstrapContext: BootstrapContext) {
    install(this)
    runningApps[this.mmid] = PromiseOut.resolve(this)

    /**
     * 对全局的自定义路由提供适配器
     * 对 nativeFetch 定义 file://xxx.dweb的解析
     */
    nativeFetchAdaptersManager.append { fromMM, request ->
      if (request.uri.scheme == "file" && request.uri.host.endsWith(".dweb")) {
        val mmid = request.uri.host
        debugFetch("DNS/nativeFetch", "$fromMM => ${request.uri}")
        val url = request.uri.toString();
        val reasonRequest =
          buildRequestX(url, InitRequest(request.method, request.headers, request.body));
        installApps[mmid]?.let {
          val (fromIpc) = connectTo(fromMM, mmid, reasonRequest)
          return@let fromIpc.request(request)
        } ?: Response(Status.BAD_GATEWAY).body(url)
      } else null
    }.removeWhen(this.onAfterShutdown)
    /** dwebDeepLink 适配器*/
    nativeFetchAdaptersManager.append { fromMM, request ->
      if (request.uri.scheme == "dweb" && request.uri.host == "") {
        debugFetch("DPLink/nativeFetch", "$fromMM => ${request.uri}")
        for (microModule in installApps) {
          for (deeplink in microModule.value.dweb_deeplinks) {
            if (request.uri.toString().startsWith(deeplink)) {
              val (fromIpc) = connectTo(fromMM, microModule.key, request)
              return@append fromIpc.request(request)
            }
          }
        }
        return@append Response(Status.BAD_GATEWAY).body(request.uri.toString())
      } else null
    }.removeWhen(this.onAfterShutdown)

    val queryAppId = Query.string().required("app_id")

    /// 定义路由功能
    apiRouting = routes(
      // 打开应用
      "/open" bind Method.GET to defineBooleanResponse { request ->
        val mmid = queryAppId(request)
        debugDNS("open/$mmid", request.uri.path)
        open(mmid)
        true
      },
      // 关闭应用
      // TODO 能否关闭一个应该应该由应用自己决定
      "/close" bind Method.GET to defineBooleanResponse { request ->
        val mmid = queryAppId(request)
        debugDNS("close/$mmid", request.uri.path)
        close(mmid)
        true
      }, "/query" bind Method.GET to defineJsonResponse { request ->
        val mmid = queryAppId(request)
        Json.encodeToString("")
        query(mmid)?.toManifest()?.toJsonElement() ?: Response(Status.OK).toJsonElement()
      }, "/observe/install-apps" bind Method.GET to defineResponse {
        val inputStream = ReadableStream(onStart = { controller ->
          val off = installApps.onChange { changes ->
            try {
              controller.enqueueBackground(
                (Json.encodeToString(
                  ChangeState(
                    changes.adds, changes.updates, changes.removes
                  )
                ) + "\n").toByteArray()
              )
            } catch (e: Exception) {
              controller.close()
              e.printStackTrace()
            }
          }
          ipc.onClose {
            off()
            controller.close()
          }
        })
        Response(Status.OK).body(inputStream)
      },
      "/observe/running-apps" bind Method.GET to defineResponse {
        val inputStream = ReadableStream(onStart = { controller ->
          val off = runningApps.onChange { changes ->
            try {
              controller.enqueueBackground(
                (Json.encodeToString(
                  ChangeState(
                    changes.adds, changes.updates, changes.removes
                  )
                ) + "\n").toByteArray()
              )
            } catch (e: Exception) {
              controller.close()
              e.printStackTrace()
            }
          }
          ipc.onClose {
            off()
            controller.close()
          }
        })
        Response(Status.OK).body(inputStream)
      })

    /// 启动 boot 模块
    connect("boot.sys.dweb").postMessage(IpcEvent.fromUtf8("activity", ""))
  }

  override suspend fun _shutdown() {
    installApps.forEach {
      it.value.shutdown()
    }
    installApps.clear()
  }

  /** 安装应用 */
  fun install(mm: MicroModule) {
    installApps[mm.mmid] = mm
  }

  /** 卸载应用 */

  fun uninstall(mmid: MMID): Boolean {
    installApps.remove(mmid)
    ioAsyncScope.launch { close(mmid) }
    return true
  }

  /** 查询应用 */
  fun query(mmid: MMID): MicroModule? {
    return installApps[mmid]
  }

  /**
   * 根据类目搜索模块
   * > 这里暂时不需要支持复合搜索，未来如果有需要另外开接口
   * @param category
   */
  fun search(category: MICRO_MODULE_CATEGORY): MutableList<MicroModule> {
    val categoryList = mutableListOf<MicroModule>()
    for (app in this.installApps.values) {
      if (app.categories.contains(category)) {
        categoryList.add(app)
      }
    }
    return categoryList
  }

  /** 打开应用 */
  private fun _open(mmid: MMID): PromiseOut<MicroModule> {
    return runningApps.getOrPut(mmid) {
      PromiseOut<MicroModule>().alsoLaunchIn(ioAsyncScope) {
        when (val openingMm = query(mmid)) {
          null -> {
            runningApps.remove(mmid)
            throw Exception("no found app: $mmid")
          }

          else -> {
            bootstrapMicroModule(openingMm)
            openingMm.onAfterShutdown {
              if (runningApps[mmid] !== null) {
                runningApps.remove(mmid)
              }
            }
            openingMm
          }
        }
      }
    }
  }

  suspend fun open(mmid: MMID): MicroModule {
    return _open(mmid).waitPromise()
  }

  /** 关闭应用 */
  suspend fun close(mmid: MMID): Int {
    return runningApps.remove(mmid)?.let { microModulePo ->
      runCatching {
        val microModule = microModulePo.waitPromise()
        microModule.shutdown()
        1
      }.getOrDefault(0)
    } ?: -1
  }
}



